File: balance-app/src/main/java/com/petro/balance/producer/BalanceChangeProducer.java
package com.petro.balance.producer;

import com.petro.balance.model.BalanceChangeEvent;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

@Slf4j
@Component
@RequiredArgsConstructor
public class BalanceChangeProducer {

    private static final String NAME = "Petro";
    private static final String SECOND_NAME = "Jack";

    private final KafkaTemplate<String, BalanceChangeEvent> kafkaTemplate;

    @Value("${spring.kafka.topics.in}")
    private final String inTopic;

    @Scheduled(fixedDelay = 20_000)
    public void publishBalanceChange() {
        sendMessage(NAME);
        sendMessage(SECOND_NAME);
    }

    private void sendMessage(String name) {
        kafkaTemplate.send(inTopic, new BalanceChangeEvent(name)).whenCompleteAsync((event, e) -> {
            if (e == null) {
                log.debug("Published event {} successfully for {}", event, NAME);
            } else {
                log.error("Failed to send balance update message for {}", NAME, e);
            }
        });
    }
}

File: balance-app-kotlin/src/main/kotlin/com/petro/balance/producer/BalanceChangeProducer.kt
package com.petro.balance.producer

import com.petro.balance.model.BalanceChangeEvent
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.beans.factory.annotation.Value
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.kafka.support.SendResult
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component

@Component
class BalanceChangeProducer(
    private val kafkaTemplate: KafkaTemplate<String, BalanceChangeEvent>,
    @Value("\${spring.kafka.topics.in}") private val inTopic: String,
) {
    @Scheduled(fixedDelay = 20000)
    fun publishBalanceChange() {
        sendMessage(NAME)
        sendMessage(SECOND_NAME)
    }

    private fun sendMessage(name: String) {
        kafkaTemplate.send(inTopic, BalanceChangeEvent(name))
            .whenCompleteAsync { event: SendResult<String, BalanceChangeEvent>?, e: Throwable? ->
                if (e == null) {
                    log.debug { "Published event $event successfully for $NAME" }
                } else {
                    log.error(e) { "Failed to send balance update message for $NAME" }
                }
            }
    }

    companion object {
        private const val NAME = "Petro"
        private const val SECOND_NAME = "Jack"
        private val log = KotlinLogging.logger {}
    }
}
